confluent_kafka/consumer.py

```diff
@@ -0,0 +1,145 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from .cimpl import Consumer as _impl
+from warnings import warn
+
+
+class Consumer(_impl):
+    """
+        Create a new Kafka Consumer instance.
+
+        To avoid spontaneous calls from non-Python threads all callbacks will only be served upon
+        calling ```client.poll()``` or ```client.flush()```.
+
+        :param dict conf: Configuration properties.
+            See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md for more information.
+        :param func key_deserializer(topic, key): Converts message key bytes to object.
+            **note** deserializers are responsible for handling NULL keys
+        :param func value_deserializer(topic, value): Converts message value bytes to object.
+            **note** deserializers are responsible for handling NULL values
+        :param func on_commit(err, [partitions]): Callback used to indicate success or failure
+            of an offset commit.
+        :param func stats_cb(json_str): Callback for statistics emitted every ``statistics.interval.ms``.
+            See https://github.com/edenhill/librdkafka/wiki/Statistics” for more information.
+        :param func throttle_cb(confluent_kafka.ThrottleEvent): Callback for throttled request reporting.
+        :param logging.handlers logger: Forwards logs from the Kafka client to the provided handler instance.
+            Log messages will only be forwarded when ``client.poll()`` or ``producer.flush()`` are called.
+        :raises TypeError: If conf is not a dict.
+    """
+
+    __slots__ = ["_key_deserializer", "_value_deserializer"]
+
+    # conf must remain optional as long as kwargs are supported
+    def __init__(self, conf={}, key_deserializer=None, value_deserializer=None,
+                 on_commit=None, stats_cb=None, throttle_cb=None, logger=None, **kwargs):
+
+        if not isinstance(conf, dict):
+            raise TypeError("expected configuration dict")
+
+        if kwargs:
+            # Handle kwargs for backwards compatibility
+            conf.update(kwargs)
+            warn("The use of kwargs is being deprecated. "
+                 "In future releases `conf` will be mandatory and "
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        self._key_deserializer = key_deserializer
+        self._value_deserializer = value_deserializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Handle both cases prior to passing along to _impl
+        # If callbacks are configured in both places parameter values take precedence.
```

Comment:
Is there a benefit of allowing cb's to be specified in more than one place?
Also, I think we might end up regretting it if we allow non-string values in the config.

Reply:
I think it is fine to enforce the callback being set in only one place, any other use is ambigious.
Let's follow Matt's rule of least surprise: enforce the callback is only specified in one place.
I don't consider that a breaking API change, it will only break bad apps.
It is literally just a couple of lines of code, something like:
```
  for var, name in [(logger, 'logger', on_commit, 'on_commit', ...)]:
     if var is not None and conf.get(name, None) is not None:
       raise ...
```

Reply:
Bad apps based off the fact we provided an API which allowed for it; sins of our fathers and what not. The amount of code is immaterial; the fact is we provided an API which allowed this. Aren't we under some obligation to deprecate this behavior responsibly? If you two think it's best then sure but this seems like the exact sort of thing that make API design so hard; you have to live with the decisions you made in the past for a long time.

Reply:
The previous behaviour was undefined and no priority was documented, fixing that is not an API breaking change since there were no promises made and an application that relies on both being specified is clearly wrong.